File: 1000 and 1 tasks/new-programming-skills/set-matrix-zeroes.cs
public class Solution {
    // сразу перекрашивать и перескакивать
    public void SetZeroes(int[][] matrix) {
        var X = matrix.Length;
        var Y = matrix[0].Length;

        var lines = new bool[X];
        var cols = new bool[Y];

        for (int i = 0; i < matrix.Length; i++) {
            for (int j = 0; j < matrix[0].Length; j++) {
                if (matrix[i][j] == 0) {
                    lines[i] = true;
                    cols[j] = true;
                }
            }
        }

        for (int i = 0; i < matrix.Length; i++) {
            for (int j = 0; j < matrix[0].Length; j++) {
                if (cols[j] || lines[i]) {
                    matrix[i][j] = 0;
                }
            }
        }
    }
}

File: 1000 and 1 tasks/new-top-150/set-matrix-zeroes.java
public class Solution {
    public void setZeroes(int[][] matrix) {
        var prevLineHasZero = false;
        for (var i = 0; i < matrix.length; i++) {

            var thisLineHasZero = false;
            for (var j = 0; j < matrix[0].length; j++) {
                if (matrix[i][j] == 0) {
                    thisLineHasZero = true;

                    // один while все равно напрягает
                    // но нельзя тормозить по нулю, потому что 0 может быть получен из-за строки
                    var itop = i - 1;
                    while (itop >= 0) {
                        matrix[itop--][j] = 0;
                    }
                }
                if (i > 0 && matrix[i-1][j] == 0) {
                    matrix[i][j] = 0;
                }
            }
            if (prevLineHasZero) {
                for (var j = 0; j < matrix[0].length; j++) {
                    matrix[i-1][j] = 0;
                }
            }

            prevLineHasZero = thisLineHasZero;
        }
        if (prevLineHasZero) {
            for (var j = 0; j < matrix[0].length; j++) {
                matrix[matrix.length-1][j] = 0;
            }
        }
    }
}